### ISSUE TITLE
WeakReferenceMessenger message received multiple times in transient pages

### ISSUE THREAD

**Issue by PureWeen** [CommunityToolkit/dotnet#657]

**Comment by roughiain** [comment]
Doing an experiment to see if the messenger class would work in a Blazor server project I came across this. 

Background service sending a text message every 15sec.

```sharp
public class TestService : BackgroundService
{
    readonly PeriodicTimer _periodicTimer = new(TimeSpan.FromSeconds(15));

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        while (await _periodicTimer.WaitForNextTickAsync(stoppingToken) && !stoppingToken.IsCancellationRequested)
        {
            await Task.Delay(10, stoppingToken);
            WeakReferenceMessenger.Default.Send(new TestMessage(DateTime.Now.ToLongDateString()));
            Console.WriteLine("MessageSent");
        }
    }
}
```

```csharp
  string message { get; set; } = string.Empty;

    protected override void OnInitialized()
    {
        WeakReferenceMessenger.Default.Register<TestMessage>(this, (r, m) =>
        {
            InvokeAsync(() =>
            {
                Console.WriteLine("Received");
                message = m.Value;
                StateHasChanged();
            });
        });
    }
```

output
```
warn: Microsoft.AspNetCore.HttpsPolicy.HttpsRedirectionMiddleware[3]
      Failed to determine the https port for redirect.
Received
Received
MessageSent
Received
Received
MessageSent
```

but if I implement `@implements IDisposable` on the page.


```csharp
     public void Dispose()
    {
        WeakReferenceMessenger.Default.Unregister<TestMessage>(this);
    }
```

the messages no longer get sent twice.

Output with IDisposable
```
warn: Microsoft.AspNetCore.HttpsPolicy.HttpsRedirectionMiddleware[3]
      Failed to determine the https port for redirect.
Received
MessageSent
Received
MessageSent
Received
MessageSent
```

not sure what is going on behind the scenes but hope this helps

**Comment by needledrag** [comment]
> Doing an experiment to see if the messenger class would work in a Blazor server project I came across this.
> 
> Background service sending a text message every 15sec.
> 
> ```
> public class TestService : BackgroundService
> {
>     readonly PeriodicTimer _periodicTimer = new(TimeSpan.FromSeconds(15));
> 
>     protected override async Task ExecuteAsync(CancellationToken stoppingToken)
>     {
>         while (await _periodicTimer.WaitForNextTickAsync(stoppingToken) && !stoppingToken.IsCancellationRequested)
>         {
>             await Task.Delay(10, stoppingToken);
>             WeakReferenceMessenger.Default.Send(new TestMessage(DateTime.Now.ToLongDateString()));
>             Console.WriteLine("MessageSent");
>         }
>     }
> }
> ```
> 
> ```cs
>   string message { get; set; } = string.Empty;
> 
>     protected override void OnInitialized()
>     {
>         WeakReferenceMessenger.Default.Register<TestMessage>(this, (r, m) =>
>         {
>             InvokeAsync(() =>
>             {
>                 Console.WriteLine("Received");
>                 message = m.Value;
>                 StateHasChanged();
>             });
>         });
>     }
> ```
> 
> output
> 
> ```
> warn: Microsoft.AspNetCore.HttpsPolicy.HttpsRedirectionMiddleware[3]
>       Failed to determine the https port for redirect.
> Received
> Received
> MessageSent
> Received
> Received
> MessageSent
> ```
> 
> but if I implement `@implements IDisposable` on the page.
> 
> ```cs
>      public void Dispose()
>     {
>         WeakReferenceMessenger.Default.Unregister<TestMessage>(this);
>     }
> ```
> 
> the messages no longer get sent twice.
> 
> Output with IDisposable
> 
> ```
> warn: Microsoft.AspNetCore.HttpsPolicy.HttpsRedirectionMiddleware[3]
>       Failed to determine the https port for redirect.
> Received
> MessageSent
> Received
> MessageSent
> Received
> MessageSent
> ```
> 
> not sure what is going on behind the scenes but hope this helps
> 
> https://github.com/roughiain/TestMvvmMessengerOnBlazorServer

Yeah I think this is what's happening, classes marked as transient aren't being disposed of when popped off the page stack so the reference to the subscription remains.  Ill see if implementing IDisposable on the view model has any effect.

Thanks

**Comment by shingming** [comment]
Same here. I use the WeakReferenceMessenger in "DetailPageViewModel" which is set as a transient service.
When I go to DetailPage (that will create an instance of "DetailPageViewModel"), return to previous page and go to DetailPage again, I will receive two times of the message.

I analyze memory management, and I see that the instance of transient service is not free even I return to previous page ( therefore, I think, the instance contains WeakReferenceMessenger which will continue to receive the message later )

![1](https://user-images.githubusercontent.com/65477270/233097065-c1f24072-0b28-4bba-a8de-7b6c6847f57c.png)

I post the question about the timing of freeing transient service.
Link: https://stackoverflow.com/questions/75642314/memory-management-in-net-maui
But, people say that it is the general case for not immediately freeing the transient service (not a problem).

In this case, I think the WeakReferenceMessenger needs to handle the condition when the transient service instance is no longer in use (even if the instance is not deleted)

**Comment by needledrag** [comment]
I agree, even if the GC hasn't kicked in to release the memory, the messages should still be unsubscribed.
The issue is you can't unsubscribe from the message in the view model once you've left the page as you don't have access to that particular instance.

I've tried unsubscribing as soon as the message is received and subscribing again at the point it's needed which works somewhat bit I've seen some duplicate messages slip through.

**Comment by KiyanYang** [comment]
I think there should be some handling of message subscriptions when leaving the page.

In ViewModel ovrride `ObservableRecipient.OnActivated()` like:

https://github.com/CommunityToolkit/MVVM-Samples/blob/8963a8720dc783658e764e22629598509e044d7f/samples/MvvmSample.Core/ViewModels/MessengerPageViewModel.cs#L72-L75

In Page, control the message state, like:

https://github.com/CommunityToolkit/MVVM-Samples/blob/8963a8720dc783658e764e22629598509e044d7f/samples/MvvmSampleUwp/Views/MessengerSendPage.xaml.cs#L27-L38

**Comment by needledrag** [comment]
Maybe I'm missing something, if you leave the page to navigate 'forward' by pushing a page on the stack and then send a message from that page, the page you navigated from wouldn't receive it?

**Comment by sergeykonar** [comment]
The issue is still there. Any ideas how it can be fixed?

**Comment by RakitinIbis** [comment]
The workaround can be to subscribe after awaiting navigation (App.Current.MainPage.Navigation.PushAsync) and unsubscribe all OnDisappearing

**Comment by ShariatPanah** [comment]
guys this problem still occurs, is there gonna be a fix for this bug?

**Comment by needledrag** [comment]
The original issue was that if you left a page that had subscribed to the message, the page (registered as Transient) would still receive the message as it was somehow still in memory.  
I ended up only subscribing right before launching the page that could send a message and unsubscribing as soon as the message was received.  In my case the subscription could only be sent once per page load.  Haven't seen the problem after this.

**Comment by ShariatPanah** [comment]
@Sergio0694 @needledrag I ended up overriding OnNavigating method in AppShell.xaml.cs and determining the source page i'm navigating from, at that point i completely clear the Messenger Queue by calling Reset on it, and then in my page i subscribe to the event. this process repeats every time i close the page and open it again. works so well for now.

**Comment by nicolasmab** [comment]
This workaround is to make sure that only the latest instance of the Page or ViewModel that is receiving the message, executes your code. If your use case is similar to the original bug report. It will not get rid of the unwanted old instances, just ignore the message when received from those ones.

The idea is to track the latest instance of your transient Page or ViewModel and send the tracker info in the message. And when receiving that message, do nothing if the instance doesn't match the tracker.


For example, my use case is:
- having transient Page1 with ViewModel1,
- navigate to Page2 with ViewModel2
- Send message from Page2
- Receive message in ViewModel1 and do something

Problem: when using the app and ending up with multiple instances of ViewModel1, the message was received in all those instances.

Solution
Add a instance tracker in your ViewModel1
```
public class ViewModel1 : BaseViewModel, IRecipient<MyMessage>
{
    public long ViewModelInstanceTracker { get; set; } = DateTime.Now.Ticks;
}
```

Send the tracker when navigating from Page1 to Page2
```
await Shell.Current.GoToAsync($"{nameof(Page2)}",
    new ShellNavigationQueryParameters { { "ViewModelInstanceTracker", viewModel.ViewModelInstanceTracker } });
```

ViewModel2 then sends the Message using the tracker
```
[QueryProperty(nameof(ViewModelInstanceTracker), "ViewModelInstanceTracker")]
public partial class Page2 : ContentPage
{
    public long ViewModelInstanceTracker { get; set; }

    private void SendMessage()
    {
        WeakReferenceMessenger.Default.Send(new MyMessage(ViewModelInstanceTracker));
    }
}
```

ViewModel1 receives the message
```
public void Receive(MyMessage message)
{
    Console.WriteLine(($"Received tracker: {message.Value}, current tracker: {ViewModelInstanceTracker}");
    if (message.Value == ViewModelInstanceTracker)
    {
        // Do something only if in latest instance
    }
}
```

**Comment by itonx** [comment]
I just faced the same scenario. After thinking about it, I don't think this is a bug in the CommunityToolkit itself. I previously found a very similar issue using Prism for WPF, so at this point I’m convinced this is more about how we combine Dependency Injection with event/message registration.

A transient service creates a new instance every time it is requested. If an event or message handler is registered in the constructor or any other initializer method, you also need to ensure it is unregistered when the instance is disposed. Most of the time, ViewModels are not disposed explicitly, so you end up depending on the GC to collect old instances. During that window, an old instance can still process messages/events, which may result in duplicated executions like this one.

**Comment by ShariatPanah** [comment]
I overcame this (so called problem) long ago, by implementing a CleanupRequesting Mechanism in ViewModels, so I'm sharing my approach with others, hoping it would help someone else in this matter.
In this approach you declare a method called RequestCleanup which Unregsiters all messages that have been Registered by current class (in our case, ViewModel), and simply call this method in BackButtonPressed (override it in page.xaml.cs) and also wherever you navigate back from current page. for a demonstration:

```
public void RequestCleanup()
{
    WeakReferenceMessenger.Default.Unregister<ClientMessage>(this);
    OnCleanupRequested?.Invoke(this, EventArgs.Empty);
    // also do whatever cleanup you need here.
}
```

call RequestCleanup, in GoBack method and also in OnBackButtonPressed:
```
private async Task GoBack()
{
    RequestCleanup();
    await Shell.Current.GoToAsync("..");
}
```

override in Page's code behind:
```
protected override bool OnBackButtonPressed()
{
    _viewModel.RequestCleanup();
    return base.OnBackButtonPressed();
}
```

this approach is working just fine for me so far.
I also created a post about this approach in my LinkedIn page, I'll include it as soon as i can
